File: src/main/java/hu/kirdev/schpincer/web/LoginController.java
package hu.kirdev.schpincer.web;

import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;

import javax.servlet.http.HttpServletRequest;

import hu.gerviba.authsch.struct.Entrant;
import hu.kirdev.schpincer.model.*;
import hu.kirdev.schpincer.service.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

import hu.gerviba.authsch.AuthSchAPI;
import hu.gerviba.authsch.response.AuthResponse;
import hu.gerviba.authsch.response.ProfileDataResponse;
import hu.gerviba.authsch.struct.Scope;
import io.swagger.annotations.Api;
import io.swagger.annotations.ApiOperation;

@Controller
public class LoginController {

    @Autowired
    AuthSchAPI authSch;

    @Autowired
    UserService users;
    
    @ApiOperation("Login re-entry point")
    @GetMapping("/loggedin")
    public String loggedIn(@RequestParam String code, @RequestParam String state, HttpServletRequest request) {
        if (!buildUniqueState(request).equals(state))
            return "index?invalid-state";
        
        Authentication auth = null;
        try {
            AuthResponse response = authSch.validateAuthentication(code);
            ProfileDataResponse profile = authSch.getProfile(response.getAccessToken());
            UserEntity user;

            if (users.exists(profile.getInternalId().toString())) {
                user = users.getById(profile.getInternalId().toString());
            	request.getSession().setAttribute("user", user);
                auth = new UsernamePasswordAuthenticationToken(code, state, getAuthorities(user));

                CardType card = cardTypeLookup(profile);
                if (user.getCardType() != card) {
                    user.setCardType(card);
                    users.save(user);
                }
            } else {
                // FIXME: old constructor internalId, fullName, email
                CardType card = cardTypeLookup(profile);
            	user = new UserEntity(profile.getInternalId().toString(),
                        profile.getSurname() + " " + profile.getGivenName(),
                        profile.getMail(),
                        "",
                        false,
                        card,
                        Collections.emptySet(),
                        1);
                users.save(user);
                auth = new UsernamePasswordAuthenticationToken(code, state, getAuthorities(user));
                request.getSession().setAttribute("user", user);
            }


            SecurityContextHolder.getContext().setAuthentication(auth);
            
        } catch (Exception e) {
            if(auth != null)
                auth.setAuthenticated(false);
            e.printStackTrace();
        }
        
        return (auth != null && auth.isAuthenticated()) ? "redirect:/" : "redirect:/?error";
    }

    private CardType cardTypeLookup(ProfileDataResponse profile) {
        CardType card = CardType.DO;
        for (Entrant entrant : profile.getEntrants()) {
            if (entrant.getEntrantType().equalsIgnoreCase("KB") && card.ordinal() < CardType.KB.ordinal()) {
                card = CardType.KB;
            } else if (entrant.getEntrantType().matches("^[ÁáAa][Bb]$") && card.ordinal() < CardType.AB.ordinal()) {
                card = CardType.AB;
            }
        }
        return card;
    }

    private List<GrantedAuthority> getAuthorities(UserEntity user) {
        List<GrantedAuthority> authorities = new ArrayList<>();
        authorities.add(new SimpleGrantedAuthority("ROLE_USER"));
        if (user.getSysadmin()) {
            authorities.add(new SimpleGrantedAuthority("ROLE_ADMIN"));
            authorities.add(new SimpleGrantedAuthority("ROLE_LEADER"));
        }
        if (user.getPermissions().contains("ROLE_LEADER"))
            authorities.add(new SimpleGrantedAuthority("ROLE_LEADER"));
        
        return authorities;
    }

    @ApiOperation("Redirection to the auth provider")
    @GetMapping("/login")
    public String items(HttpServletRequest request) {
        return "redirect:" + authSch.generateLoginUrl(buildUniqueState(request),
                Scope.BASIC, Scope.GIVEN_NAME, Scope.SURNAME, Scope.MAIL, Scope.ENTRANTS);
    }

    static String buildUniqueState(HttpServletRequest request) {
        return hashString(request.getSession().getId()
                + request.getLocalAddr()
                + request.getLocalPort());
    }

    static final String hashString(String in) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            return String.format("%064x", new BigInteger(1, digest.digest(in.getBytes())));
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
        }
        return "error";
    }

    @ApiOperation("Logout user")
    @GetMapping("/logout")
    public String logout(HttpServletRequest request) {
    	request.removeAttribute("user");
    	request.getSession().removeAttribute("user");
    	request.changeSessionId();
    	return "redirect:/?logged-out";
    }
    
}

File: src/main/kotlin/hu/kirdev/schpincer/web/LoginController.kt
package hu.kirdev.schpincer.web

import hu.gerviba.authsch.AuthSchAPI
import hu.gerviba.authsch.response.ProfileDataResponse
import hu.gerviba.authsch.struct.Scope
import hu.kirdev.schpincer.config.Role
import hu.kirdev.schpincer.model.CardType
import hu.kirdev.schpincer.model.UserEntity
import hu.kirdev.schpincer.service.CircleService
import hu.kirdev.schpincer.service.UserService
import io.swagger.annotations.ApiOperation
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import java.util.*
import javax.servlet.http.HttpServletRequest

const val USER_SESSION_ATTRIBUTE_NAME = "user_id"
const val USER_ENTITY_DTO_SESSION_ATTRIBUTE_NAME = "user"
const val CIRCLE_OWNERSHIP_SESSION_ATTRIBUTE_NAME = "circles"

@Controller
open class LoginController {

    @Autowired
    private lateinit var authSch: AuthSchAPI

    @Autowired
    private lateinit var users: UserService

    @Autowired
    private lateinit var circles: CircleService

    @Value("\${schpincer.sysadmins:}")
    private lateinit var systemAdmins: String

    @ApiOperation("Login re-entry point")
    @GetMapping("/loggedin")
    fun loggedIn(@RequestParam code: String, @RequestParam state: String, request: HttpServletRequest): String {
        if (buildUniqueState(request) != state)
            return "index?invalid-state"

        var auth: Authentication? = null
        try {
            val response = authSch.validateAuthentication(code)
            val profile = authSch.getProfile(response.accessToken)
            val user: UserEntity
            val ownedCircles = getOwnedCircleIds(profile)
            if (users.exists(profile.internalId.toString())) {
                user = users.getById(profile.internalId.toString())
                val card = cardTypeLookup(profile)
                if (user.cardType !== card) {
                    user.cardType = card
                    users.save(user)
                }
                val permissionsByVIR = getCirclePermissionList(ownedCircles)
                if (!user.permissions.containsAll(permissionsByVIR)) {
                    permissionsByVIR.addAll(user.permissions)
                    user.permissions = permissionsByVIR
                    users.save(user)
                }
            } else {
                val card = cardTypeLookup(profile)
                user = UserEntity(profile.internalId.toString(),
                        profile.surname + " " + profile.givenName,
                        profile.mail,
                        "",
                        false,
                        card, getCirclePermissionList(ownedCircles),
                        1)
                users.save(user)
            }

            if (systemAdmins.split(",").contains(user.uid))
                user.sysadmin = true

            auth = UsernamePasswordAuthenticationToken(code, state, getAuthorities(user))

            request.session.setAttribute(USER_SESSION_ATTRIBUTE_NAME, user.uid)
            request.session.setAttribute(USER_ENTITY_DTO_SESSION_ATTRIBUTE_NAME, user)
            request.session.setAttribute(CIRCLE_OWNERSHIP_SESSION_ATTRIBUTE_NAME, ownedCircles)
            SecurityContextHolder.getContext().authentication = auth
        } catch (e: Exception) {
            auth?.isAuthenticated = false
            e.printStackTrace()
        }
        return if (auth != null && auth.isAuthenticated) "redirect:/" else "redirect:/?error"
    }

    private fun getOwnedCircleIds(profile: ProfileDataResponse): List<Long> {
        return profile.eduPersonEntitlements
                .filter { it.status == "körvezető" }
                .mapNotNull { circles.findByVirGroupId(it.id)?.id }
    }

    private fun getCirclePermissionList(circles: List<Long>): MutableSet<String> {
        val permissions = mutableSetOf<String>()
        if (circles.isNotEmpty()) {
            permissions.add("ROLE_LEADER")
            permissions.addAll(circles.map { "CIRCLE_${it}" })
        }
        return permissions
    }

    private fun cardTypeLookup(profile: ProfileDataResponse): CardType {
        var card = CardType.DO
        for (entrant in profile.entrants) {
            if (entrant.entrantType.equals("KB", ignoreCase = true) && card.ordinal < CardType.KB.ordinal) {
                card = CardType.KB
            } else if (entrant.entrantType.matches("^[ÁáAa][Bb]$".toRegex()) && card.ordinal < CardType.AB.ordinal) {
                card = CardType.AB
            }
        }
        return card
    }

    private fun getAuthorities(user: UserEntity): List<GrantedAuthority> {
        val authorities: MutableList<GrantedAuthority> = ArrayList()
        authorities.add(SimpleGrantedAuthority("ROLE_${Role.USER.name}"))
        if (user.sysadmin) {
            authorities.add(SimpleGrantedAuthority("ROLE_${Role.ADMIN.name}"))
            authorities.add(SimpleGrantedAuthority("ROLE_${Role.LEADER.name}"))
        }
        if (user.permissions.contains("ROLE_${Role.LEADER.name}"))
            authorities.add(SimpleGrantedAuthority("ROLE_${Role.LEADER.name}"))
        return authorities
    }

    @ApiOperation("Redirection to the auth provider")
    @GetMapping("/login")
    fun items(request: HttpServletRequest): String {
        return "redirect:" + authSch.generateLoginUrl(buildUniqueState(request),
                Scope.BASIC, Scope.GIVEN_NAME, Scope.SURNAME, Scope.MAIL, Scope.ENTRANTS, Scope.EDU_PERSON_ENTILEMENT)
    }

    fun buildUniqueState(request: HttpServletRequest): String {
        return (request.session.id
                + request.localAddr
                + request.localPort).sha256()
    }

    @ApiOperation("Logout user")
    @GetMapping("/logout")
    fun logout(request: HttpServletRequest): String {
        request.removeAttribute(USER_SESSION_ATTRIBUTE_NAME)
        request.removeAttribute(USER_ENTITY_DTO_SESSION_ATTRIBUTE_NAME)
        request.removeAttribute(CIRCLE_OWNERSHIP_SESSION_ATTRIBUTE_NAME)
        request.session.removeAttribute(USER_SESSION_ATTRIBUTE_NAME)
        request.session.removeAttribute(USER_ENTITY_DTO_SESSION_ATTRIBUTE_NAME)
        request.session.removeAttribute(CIRCLE_OWNERSHIP_SESSION_ATTRIBUTE_NAME)
        request.changeSessionId()
        return "redirect:/?logged-out"
    }

}
